### PR TITLE
Adds proof harnesses for aws_byte_cursor_advance* functions

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*
1. Updates implementation of `aws_byte_cursor_advance*` functions with pre- and post-conditions;
2. Adds proof harnesses for `aws_byte_cursor_advance*` functions;

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by feliperodri** [comment]
@tautschnig @karkhaz @angelhof could you review the proof harnesses?

**Comment by feliperodri** [comment]
@justinboswell @JonathanHenson @ColdenCullen @graebm could you review the changes in `byte_buf.h`?

File: include/aws/common/byte_buf.h [hunk 1]
```diff
@@ -605,7 +605,8 @@ AWS_STATIC_IMPL size_t aws_nospec_mask(size_t index, size_t bound) {
  * Note that if len is above (SIZE_MAX / 2), this function will also treat it as
  * a buffer overflow, and return NULL without changing *buf.
  */
-AWS_STATIC_IMPL struct aws_byte_cursor aws_byte_cursor_advance(struct aws_byte_cursor *cursor, size_t len) {
+AWS_STATIC_IMPL struct aws_byte_cursor aws_byte_cursor_advance(struct aws_byte_cursor *const cursor, const size_t len) {
```
Comment (tautschnig) [inline include/aws/common/byte_buf.h]: Shouldn't the `const` also be added to the other variant?

Comment (feliperodri) [reply include/aws/common/byte_buf.h]: Yes, done.

**Pull request merged**